src/pprof/ddprof_pprof.cc

```diff
@@ -34,54 +32,43 @@ std_string_2_slice_c_char(const std::string &str) {
 #define SLICE_LITERAL(str)                                                     \
   (struct ddprof_ffi_Slice_c_char) { .ptr = (str), .len = sizeof(str) - 1 }
 
+#define PPROF_MAX_LABELS 5
+
 static ddprof_ffi_Slice_c_char ffi_empty_char_slice(void) {
   return (struct ddprof_ffi_Slice_c_char){.ptr = (NULL), .len = 0};
 }
 
-static const struct ddprof_ffi_ValueType s_sample_count = {
-    .type_ = SLICE_LITERAL("sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_time = {
-    .type_ = SLICE_LITERAL("cpu-time"),
-    .unit = SLICE_LITERAL("nanoseconds"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_sample = {
-    .type_ = SLICE_LITERAL("cpu-sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-DDRes pprof_create_profile(DDProfPProf *pprof, const PerfOption *options,
-                           unsigned nbOptions) {
-  // Create one value
-  struct ddprof_ffi_ValueType perf_value_type[MAX_TYPE_WATCHER + 1];
-  unsigned value_index = 0;
-  perf_value_type[value_index++] = s_sample_count;
-
-  /* Add one value type per watcher */
-  for (; value_index < nbOptions + 1; ++value_index) {
-    ddprof_ffi_Slice_c_char perf_value_slice = {.ptr = options->label,
-                                                .len = strlen(options->label)};
-    ddprof_ffi_Slice_c_char perf_unit = {.ptr = options->unit,
-                                         .len = strlen(options->unit)};
-
-    perf_value_type[value_index].type_ = perf_value_slice;
-    perf_value_type[value_index].unit = perf_unit;
+DDRes pprof_create_profile(DDProfPProf *pprof, unsigned type_default,
+    int64_t periodfreq_default) {
+  // Add all the defined profile types (perf_watcher.h), skipping the first
+  struct ddprof_ffi_ValueType perf_value_type[DDPROF_PWT_LENGTH];
+  unsigned num_profiles = 0;
+  for (unsigned i = 0; i < DDPROF_PWT_LENGTH; ++i) {
```

Comment:
So for now we add all of the types as columns inside a pprof.
At aggregation we look up what is the matching index considering the index of the watcher.
In a future PR I would also consider the position of the watcher in the ddprof_context.
The index to the value inside the pprof should be more relevant to the values we store inside a pprof ?

Reply:
Changed this,
1. During setup, all of the watchers are extracted from a DDProfContext and their underlying sample types are checked
2. Only sample types which are used by at least one watcher are populated in the pprof
3. During that process, a permutation vector mapping (static) sample type IDs to (dynamic) pprof IDs is created
4. that permutation vector is used during pprof_aggregate to correctly map both the main sample_type_id as well as the count sample_type_id (if one is defined for that sample type)

There's another way of doing it, which is to add more state to the watcher.  I originally wanted to do that, but opted for the permutation strategy since to my eye indexing one list into a smaller list is simpler.  I guess this is also useful if we ever dynamically add profile types during the lifetime of a single context (then just populate the corresponding entry in the permutation vector), but I find that very unlikely in the next ~6 months...

Reply:
I would have added state to PerfWatcher by adding the index in pprof sample array, this would have avoided passing additional arguments to aggregation, but it's a matter of personal taste ;)

Reply:
Yeah, I very much disliked passing the pv around as an argument.  Although to be fair, as you observe, if you're going to pass *arrays* rather than *objects* you necessarily have to pass two arrays (one for the array, one for the index/permutation vector). Anyway, the flexibility you get from using composable objects is totally lost on this interface, which should have very little flexibility in practice. I've backed off from that technique here.

Accordingly, I've changed the PerfWatcher to contain two indices into the pprof ValueType field (one for the value, one for its count), which are populated on pprof creation.  This enables the PerfWatcher to record its original Profiling information as well as index into the pprof correctly.  Whether or not the count is used is still inferred from the original ID, even though the count is correctly populated in the pprof now if it is specified (as other than NOCOUNT).